### ISSUE TITLE
Custom Hilbert spaces

### ISSUE THREAD

**Issue by VolodyaCO** [netket/netket#398]
I am interested in creating a custom Hilbert space, where each node has a variable Hilbert size. I was expecting CustomHilbert to work something like:

```
g = nk.graph.CustomGraph([[0, 1],[1, 2],[2, 0]])
hi = nk.hilbert.CustomHilbert(graph=g, local_size=[2,3,4])
```

Now, operators acting on the first node would be represented by 2x2 matrices, the ops acting on the second node would be represented by 3x3, and so on. Apparently, `CustomHilbert` only creates hilbert spaces of size 2 for each node, and receives eigenvalues of the states, which I don't understand.

I have two questions. Is there any custom Hilbert space that works as I explain above? And, how does the current CustomHilbert exactly work?

**Comment by gcarleo** [comment]
No this would be a nice feature to have, but at the moment we don't support heterogeneous hilbert spaces unfortunately. The quick and dirty way to achieve what you want is to work with the largest local Hilbert space and use operators that effectively act as identities on the extra degrees of freedom... It's a bit ugly but it works. The other alternative would be to map everything onto a larger homogenous binary hilbert space, assuming that each local_size in your case is a multiple of 2.

**Comment by PhilipVinc** [comment]
I recently had a chat about this.
I'm quite confident that it shouldn't be too hard to implement.
I think the way to go would be to define a new Hilbert space, `CompositeHilbertSpace`, which maintains a list of sub-spaces, and do some algebra for correctly iterating the set of states.

The biggest complication would be in supporting the samplers.
Dispatching on the type of the space would be a nice thing to do. And for jax machines this would result in a zero cost once methods are jitted (aka, always).
Otherwise we should change all samplers to assume non uniform spaces.

**Comment by VolodyaCO** [comment]
I can work on this in August

**Comment by gcarleo** [comment]
This is implemented in #539 !
